### PR TITLE
Add UTF8Encoding tests

### PR DESCRIPTION
**Pull request by hughbe**
- All ASCII chars
- Valid surrogate pairs
- Invalid surrogate pairs
- Invalid unicode chars
- Invalid bytes
- Change Decode(..., char[]) to Decode(..., string) for convenience

/cc @tarekgh @jonhanna

### CODE REVIEW

**Comment by tarekgh** [comment]
LGTM. Excellent work @hughbe

File: src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            // Invalid bytes
+            byte[] validSurrogateBytes = new byte[] { 240, 144, 128, 128 };
+            Decode(validSurrogateBytes, 0, 3, "\uFFFD");
+            Decode(validSurrogateBytes, 1, 3, "\uFFFD\uFFFD\uFFFD");
+            Decode(validSurrogateBytes, 0, 2, "\uFFFD");
+            Decode(validSurrogateBytes, 1, 2, "\uFFFD\uFFFD");
+            Decode(validSurrogateBytes, 2, 2, "\uFFFD\uFFFD");
+            Decode(validSurrogateBytes, 2, 1, "\uFFFD");
+
+            Decode(new byte[] { 0xC0, 0xAF }, 0, 2, "\uFFFD\uFFFD");
```
Comment (JonHanna) [inline src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs]: Mat

Comment (JonHanna) [inline src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs]: Maybe comment that this is an example of an overlong encoding (and insecure). Give a helping hand to the hypothetical developer who was hoping to fix or optimise something and ended up triggering this.

**Comment by JonHanna** [comment]
:+1;

File: src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs]: Good idea, thanks. I added a comment and added some more tests

**Comment by tarekgh** [comment]
test Innerloop OSX Debug Build and Test please
test Innerloop OSX Release Build and Test please

**Comment by tarekgh** [comment]
test Innerloop OSX Debug Build and Test please
 test Innerloop OSX Release Build and Test please

**Comment by tarekgh** [comment]
test Innerloop OSX Debug Build and Test please

**Comment by tarekgh** [comment]
test Innerloop OSX Debug Build and Test please

**Comment by tarekgh** [comment]
test Innerloop OSX Debug Build and Test please

**Comment by tarekgh** [comment]
@dotnet-bot please check this one

**Comment by tarekgh** [comment]
test Innerloop OSX Debug Build and Test

File: src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            // Invalid bytes
+            byte[] validSurrogateBytes = new byte[] { 240, 144, 128, 128 };
+            Decode(validSurrogateBytes, 0, 3, "\uFFFD");
+            Decode(validSurrogateBytes, 1, 3, "\uFFFD\uFFFD\uFFFD");
+            Decode(validSurrogateBytes, 0, 2, "\uFFFD");
+            Decode(validSurrogateBytes, 1, 2, "\uFFFD\uFFFD");
+            Decode(validSurrogateBytes, 2, 2, "\uFFFD\uFFFD");
+            Decode(validSurrogateBytes, 2, 1, "\uFFFD");
+
+            // These are examples of overlong sequences - i.e. an invalid sequence
```
Comment (JonHanna) [inline src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs]: Nit: "i.e. an invalid sequence" could be read as meaning it applies to all invalid sequences.

**Comment by JonHanna** [comment]
Test Innerloop OSX Debug Build and Test 
(Build timeout).

File: src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs]: Yeah I left it in accidently when pluralising the comment! Will fix

Comment (hughbe) [inline src/System.Text.Encoding/tests/UTF8Encoding/UTF8EncodingDecode.cs]: fixed

**Comment by tarekgh** [comment]
test Innerloop OSX Release Build and Test please

**Pull request merged**